README.md
# copyseparator
Assembling Long Gene Copies from Short Read Data

install.packages("copyseparator")


NEWS.md
3# copyseparator 1.0.0

* Added a `NEWS.md` file to track changes to the package.


R/toydata.R
#' @title toydata
#'
#' Provides small example dataset for major functions. They are located at: /inst/extdata
#'
#' The file "toydata.fasta" is for functions "copy_separate" and "subset_downsize".
#' The file "combined_con.txt" is for the function "copy_assemble".
#' The file "toysubset.fasta" is for the function "copy_detect".
#' The file "Final_two_copiess.fasta" is for the function "copy_validate".
#'
#' @name toydata
#' @docType data
#' @format fasta
#' @references The data was collected by the author and is unpublished.
#' @keywords dataset
#' @examples
#'

## copy_separate
copy_separate("inst/extdata/toydata.fasta",2,300,225,1)

## copy_assemble
copy_assemble("inst/extdata/combined_con.fasta",2,1)

## copy_validate
copy_validate("inst/extdata/Final_two_copies.fasta",2,300,1)

## subset_downsize
subset_downsize("inst/extdata/toydata.fasta", 300,225,1)

## copy_detect
copy_detect("inst/extdata/toysubset.fasta",2,1)



## delete all resulting files after testing
unlink("*.fasta")
unlink("*.txt")
unlink("*.pdf")
unlink("inst/extdata/toydata_intermediate_files", recursive = TRUE)
unlink("inst/extdata/toydata_combined*.txt")
unlink("inst/extdata/toysubset_*")






R/copy_detect.R
#' @title copy_detect
#'
#' @description Separates two or more gene copies from a single subset of short reads.
#'
#' @param filename A fasta file contains short reads from a single subset generated by "subset_downsize".
#'
#' @param copy_number An integer (e.g. 2,3, or 4) giving the anticipated number of gene copies in the input file.
#'
#' @param verbose Turn on (verbose=1; default) or turn off (verbose=0) the output.
#'
#' @return A fasta alignment of the anticipated number of gene copies.
#'
#' @importFrom seqinr read.fasta write.fasta
#'
#' @importFrom stringr str_sort
#'
#' @importFrom kmer otu
#'
#' @importFrom Biostrings readDNAStringSet
#'
#' @importFrom ape as.character.DNAbin read.FASTA
#'
#' @importFrom DECIPHER ConsensusSequence
#'
#' @importFrom beepr beep
#'
#' @export copy_detect
#'

copy_detect<-function(filename,copy_number, verbose=1)
{
  sink("log.txt", append=FALSE, split=TRUE) # begin to record log
  error_log_function <- function() {
    cat(geterrmessage(), file="Error_log.txt", append=T)
  }

  if (copy_number<=1) stop ("The anticipated copy number must be a number larger than one!")

    Sub_set <- ape::as.character.DNAbin(ape::read.FASTA(file=filename, type = "DNA"))
    if (verbose) { cat(paste0("Clustering analyses for ", filename,"\n"))}

    filename_short <- gsub("[:.:].*","", filename) # remove file extensions, e.g. ".fasta", ".txt"

    # find the threshold range for OTU to find the major clusters (number=copy_number) for each subset
    for (m in seq(0.3,1, by = 0.1)) {
      Subset_OTU <- kmer::otu(Sub_set, k = 5, threshold = m, method = "central", nstart = 20)
      if (verbose) { cat(paste0("threshold = ",m),"\n")}
      if (verbose) { cat(unique(Subset_OTU),"\n")}
      if (length(unique(Subset_OTU))>=copy_number) {break}
    }

    # try different threshold values in the range found above
    for (i in seq(m-0.09,m, by = 0.01)) {
      Subset_OTU <- kmer::otu(Sub_set, k = 5, threshold = i, method = "central", nstart = 20)
      if (verbose) { cat(paste0("threshold = ",i),"\n")}
      if (verbose) { cat(unique(Subset_OTU),"\n")}
      if (length(unique(Subset_OTU))>=copy_number) {break}
    }

    reads_each_cluster <- sapply(unique(Subset_OTU), function(x) length(which(Subset_OTU==x)))

    if (verbose) { cat(paste0("Best threshold found = ",i),"\n")}
    if (verbose) { cat(unique(Subset_OTU),"\n")}
    if (verbose) { cat("Number of reads in each cluster\n")}
    if (verbose) { cat(reads_each_cluster,"\n")}

    for (j in (1:copy_number)) {
      Picked_cluster <- Sub_set[which(Subset_OTU==unique(Subset_OTU)[which(reads_each_cluster==sort(reads_each_cluster)[length(unique(Subset_OTU))-j+1])])]
      seqinr::write.fasta(sequences = Picked_cluster,
                names = labels(Picked_cluster),
                file.out = paste0(filename_short,"_cluster_",j,".fasta"))
      if (verbose) { cat(paste0("Number of reads in picked cluster ",j, " = ", length(Picked_cluster),"\n"))}

    # calcuate the consensus sequence for the clusters of the subset
    seqinr::write.fasta(sequences = as.character(DECIPHER::ConsensusSequence(Biostrings::readDNAStringSet(paste0(filename_short,"_cluster_",j,".fasta"),format="fasta",nrec=-1L, skip=0L),threshold = 0.4,
    ambiguity = TRUE, noConsensusChar = "N")[1]),names = paste0(filename_short,"_cluster_",j,"_consensus"), file.out = paste0(filename_short,"_cluster_",j,"_consensus.fasta"))
    }

    # put together all the consensus sequences into one file
    All_consensus <- lapply(1:copy_number, function (x) seqinr::read.fasta(file = paste0(filename_short,"_cluster_",x,"_consensus.fasta"), seqtype = "DNA",
                                                                                      as.string = TRUE,forceDNAtolower = FALSE,set.attributes = FALSE, whole.header = TRUE))

    seqinr::write.fasta(sequences=All_consensus, names=names(as.data.frame(All_consensus)), file.out=paste0(filename_short,"_consensus_list.fasta"))

    cat("Run finished!\n")
    beepr::beep(sound = 1, expr = NULL) # make a sound when run finishes
    options("error" = error_log_function)
    sink() # turn off log
}




R/example_data.R
#' toydata
#' 
#' Provides small example dataset for major functions. They are located at: /inst/extdata
#'
#' The file "toydata.fasta" is for functions "copy_separate" and "subset_downsize".
#' The file "combined_con.txt" is for the function "copy_assemble".
#' The file "toysubset.fasta" is for the function "copy_detect".
#' The file "Final_two_copiess.fasta" is for the function "copy_validate".
#'
#' @docType data
#' @usage
   # copy_separate("inst/extdata/toydata.fasta",2,300,225,1)
   # copy_assemble("inst/extdata/combined_con.fasta",2,1)
   # copy_validate("inst/extdata/Final_two_copies.fasta",2,300,1)
   # subset_downsize("inst/extdata/toydata.fasta",300,225,1)
   # copy_detect("inst/extdata/toysubset.fasta",2,1)

### After the test runs, you may run the following codes to delete all the resulting files.
   # unlink("*.fasta")
   # unlink("*.txt")
   # unlink("*.pdf")
   # unlink("inst/extdata/toydata_intermediate_files", recursive = TRUE)
   # unlink("inst/extdata/toydata_combined*.txt")
   # unlink("inst/extdata/toysubset_*")

#' @format fasta
#' @references This data set was collected by the author and is unpublished.
#' @keywords dataset


R/copy_assemble.R
#' @title copy_assemble
#'
#' @description Assembles a small number of overlapping DNA sequences into their respective gene copies.
#'
#' @param filename A fasta alignment of a small number of overlapping DNA sequences (results from "copy_separate") covering the entire length of the target gene. Check the alignment carefully before proceeding.
#'
#' @param copy_number An integer (e.g. 2,3, or 4) giving the anticipated number of gene copies. Must be the same value as used for "copy_separate".
#'
#' @param verbose Turn on (verbose=1; default) or turn off (verbose=0) the output.
#'
#' @return A fasta alignment of the anticipated number of full-length gene copies.
#'
#' @importFrom seqinr read.fasta write.fasta
#'
#' @importFrom stringr str_count str_order
#'
#' @importFrom Biostrings readDNAStringSet
#'
#' @importFrom DECIPHER ConsensusSequence
#'
#' @importFrom beepr beep
#'
#' @export copy_assemble
#'

copy_assemble<-function(filename,copy_number, verbose=1)
{
  sink("log.txt", append=FALSE, split=TRUE) # begin to record log
  error_log_function <- function() {
    cat(geterrmessage(), file="Error_log.txt", append=T)
  }

  if (copy_number<=1) stop ("The anticipated copy number must be a number larger than one!")

  Consensus_seq <- seqinr::read.fasta(file = filename,seqtype = "DNA", as.string = TRUE,forceDNAtolower = FALSE,set.attributes = FALSE)

  subset_sum <- length(Consensus_seq)/copy_number

  seq_to_con <- character(0)
  for (l in 1:copy_number) {
    for (i in seq(l, (subset_sum-1)*copy_number, copy_number)){
      ambiguity_num <- character(0)
      for (j in 1:copy_number) {
        seqinr::write.fasta(sequences = c(Consensus_seq[i], Consensus_seq[i+copy_number+j-l]), names(c(Consensus_seq[i], Consensus_seq[i+copy_number])),file.out = paste0("cnx_",i,"_",i+copy_number+j-l,".fasta"))
        ambiguity_num <- append(ambiguity_num, sum(as.integer(stringr::str_count(as.character(DECIPHER::ConsensusSequence(Biostrings::readDNAStringSet(paste0("cnx_",i,"_",i+copy_number+j-l,".fasta"), format="fasta",nrec=-1L, skip=0L),
                                                                                                       threshold = 0.4,ambiguity = TRUE, noConsensusChar = "N")), c("M", "K", "R", "Y","N","W", "S", "H", "V", "D", "B")))))
        if (verbose) { cat(paste0("Matching sequences ",i, " & ", i+copy_number+j-l, "\n"))}
      }
      if (verbose) { cat("Number of ambiguous sites for each match:", ambiguity_num, "\n")}
      if (verbose) { cat(paste0("--- The pair ", which(as.numeric(ambiguity_num)==min(as.numeric(ambiguity_num))), " has fewer ambiguous sites and should be assembled together\n"))}

      seq_to_con <- append(seq_to_con, paste0("A",i,"B_A",i+copy_number+which(as.numeric(ambiguity_num)==min(as.numeric(ambiguity_num)))-l,"B"))
    }
  }
  seq_to_con1 <- seq_to_con[stringr::str_order(seq_to_con, decreasing = FALSE, na_last = TRUE, locale = "en",numeric = TRUE)] # order numerically

  unlink("cnx_*") # Delete all intermediate files whose names begin with "cnx_"

  if (verbose) { cat("*************************************************************************\n")}
  # To find out which sequences belong to which gene copy
  Copy_list <- data.frame()
  for (i in 1:copy_number) {
    cp <- c(i)
    for (j in 1:(subset_sum-copy_number)) {
      m <- which(gsub(".*_", "", seq_to_con1[cp[j]]) == gsub("_.*", "", seq_to_con1)) # From A1B_A4B to A4B_A5B, then to A5B_A8B, then to A8B_A10B ...; linking 1, 4, 5, 8, 10 ...
      cp <- append(cp,m)
    }
    cp_all <- as.numeric(gsub("[AB]","",c(gsub("_.*", "", seq_to_con1[cp]), gsub(".*_", "", seq_to_con1[cp][subset_sum-copy_number+1]))))
    Copy_list <- append(Copy_list,as.data.frame(cp_all))
  }

  seq_list <- character(0)
  for (i in 1:copy_number) {
    cat("List of sequences to be assembled for gene copy", i, ": ", Copy_list[[i]], "\n")
    for (j in (1:copy_number)[-i]){
      seq_list <- append(seq_list, intersect(as.character(Copy_list[[i]]),as.character(Copy_list[[j]])))
    }
  }

  cat("--- Sequences involved in the assembling of multiple gene copies: ", stringr::str_sort(unique(seq_list), numeric=TRUE), "\n")
  cat("Warning! If there are sequences involved in the assembling of multiple gene copies, please check your input file carefully and try to do the assembling again or do it manually!\n")


  all_copies_final <- character(0)
  for (i in 1:copy_number) {
    copy_subseqs <- lapply(as.numeric(unlist(Copy_list[i])), function(x) Consensus_seq[x])
    copy_subseqs_name <- lapply(as.numeric(unlist(Copy_list[i])), function(x) names(Consensus_seq[x]))
    seqinr::write.fasta(sequences = copy_subseqs, copy_subseqs_name,file.out=paste0("Copy_",i,"_subseqs.fasta"))
    seqinr::write.fasta(sequences = as.character(DECIPHER::ConsensusSequence(Biostrings::readDNAStringSet(paste0("Copy_",i,"_subseqs.fasta"), format="fasta",nrec=-1L, skip=0L),
                                                           threshold = 0.4,ambiguity = TRUE, noConsensusChar = "N")), paste0("Copy_",i,"_final"),file.out=paste0("Copy_",i,"_final.fasta"))
    all_copies_final <- append(all_copies_final, seqinr::read.fasta (paste0("Copy_",i,"_final.fasta"), seqtype = "DNA", as.string = TRUE,forceDNAtolower = FALSE,set.attributes = FALSE))
  }
  seqinr::write.fasta(sequences = all_copies_final, names(all_copies_final),file.out=paste0("All_final_copies.fasta"))
  unlink("*_final.fasta")

  if (verbose) { cat("*************************************************************************\n")}
  cat("Run finished!\n")
  beepr::beep(sound = 1, expr = NULL) # make a sound when run finishes
  options("error" = error_log_function)
  sink() # turn off log
}


R/copy_validate.R
#' @title copy_validate
#'
#' @description A tool to help identify incorrectly assembled chimeric sequences.
#'
#' @param filename A DNA alignment in fasta format that contains sequences of two or more gene copies (e.g. results from "copy_assemble").
#'
#' @param copy_number An integer (e.g. 2,3, or 4) giving the number of gene copies in the input file.
#'
#' @param read_length An integer (e.g. 250, or 300) giving the read length of your Next-generation Sequencing data.
#'
#' @param verbose Turn on (verbose=1; default) or turn off (verbose=0) the output.
#'
#' @return A histogram in pdf format showing the relationships between the physical distance between neighboring variable sites and read length.
#'
#' @importFrom seqinr read.fasta
#'
#' @importFrom grDevices dev.off pdf
#'
#' @importFrom graphics abline hist text
#'
#' @importFrom beepr beep
#'
#' @export copy_validate


copy_validate<-function(filename,copy_number,read_length, verbose=1)
{
  sink("log.txt", append=FALSE, split=TRUE) # begin to record log
  error_log_function <- function() {
    cat(geterrmessage(), file="Error_log.txt", append=T) # begin to record error log
  }

  all_copies <- seqinr::read.fasta(filename, seqtype = "DNA", as.string = TRUE,forceDNAtolower = FALSE,set.attributes = FALSE)

  all_dist <-  character(0)
  for (i in 1:(copy_number)) {
    for (j in i:(copy_number)) {
      Paralog_diff <- which(unlist(strsplit(all_copies[[i]], ""))!=unlist(strsplit(all_copies[[j]], "")))

      dist_var_sites <- sapply(2:length(Paralog_diff), function(x) Paralog_diff[x]-Paralog_diff[x-1])
      all_dist <- append(all_dist,  dist_var_sites)
      if (verbose) { cat(paste0("Copy ", i, " and Copy ", j, " is different at site: ", Paralog_diff,"\n"))}
    }
  }

  dist_final <- sort(as.numeric(all_dist))

  if (max(dist_final)>read_length)
   cat ("Warning! Distance between some neighboring variable sites is larger than the read length (see the plot in pdf), chimeric sequences may have formed during assembling!", "\n")

  text_to_show <- "read_length"

  grDevices::pdf(file="Distance between neighboring variable sites VS. Read length.pdf", width=8, height=8)

  graphics::hist(dist_final,main="Distance between neighboring variable sites VS. Read length", xlab="Base pairs",
       ylim=c(0,length(dist_final)/2),xlim=c(0,read_length+100),breaks="Sturges",col="blue")

  graphics::abline(v=read_length, col="red", lwd=3, lty=2)
  graphics::text(read_length, length(dist_final)/2, text_to_show, pos = 4, offset = 0.5)

  cat('Run finished! Please check results in the file: "Distance between neighboring variable sites VS. read length.pdf"',"\n")

  beepr::beep(sound = 1, expr = NULL) # make a sound when run finishes
  options("error" = error_log_function)
  sink() # turn off log
  invisible(grDevices::dev.off())
}






R/subset_downsize.R
#' @title subset_downsize
#'
#' @description Subdivides the imported read alignment into subsets and then downsizes each subset by deleting those sequences that have too many gaps or missing data.
#'
#' @param filename A fasta file contains thousands of short reads that have been mapped to a reference. The reference and reads that are not directly mapped to the reference need to be removed after mapping.
#'
#' @param read_length An integer (e.g. 250, or 300) giving the read length of your Next-generation Sequencing data. This method is designed for read length >=250bp.
#'
#' @param overlap An integer describing number of base pairs of overlap between adjacent subsets. More overlap means more subsets.
#'
#' @param verbose Turn on (verbose=1; default) or turn off (verbose=0) the output.
#'
#' @return A number of overlapping subsets (before and after downsizing) of the input alignment.
#'
#' @importFrom seqinr read.fasta write.fasta
#'
#' @importFrom stringr str_count str_sort
#'
#' @importFrom beepr beep
#'
#' @export subset_downsize
#' 

subset_downsize<-function(filename,read_length,overlap, verbose=1)
{
  sink("log.txt", append=FALSE, split=TRUE) # begin to record log
  error_log_function <- function() {
    cat(geterrmessage(), file="Error_log.txt", append=T)
  }

  if (read_length<250) warning ("This method is designed for read length 250bp or longer. Short reads can easily result in chimeric sequences.")

  NGSreads <- seqinr::read.fasta(file = filename,seqtype = "DNA", as.string = TRUE,forceDNAtolower = FALSE,set.attributes = FALSE)

  total_reads <- length(NGSreads); if (verbose) { cat(paste0("Total number of reads imported = ",total_reads,"\n"))}
  alignment_length <- nchar(NGSreads[[1]]); if (verbose) { cat(paste0("Length of the alignment = ",alignment_length,"\n"))}

  if (verbose) { cat(paste0("Read length = ",read_length,"\n"))}
  if (verbose) { cat(paste0("Overlap between adjacent subsets = ",overlap,"\n"))}
  if (overlap>=read_length) stop("Overlap between adjacent subsets must be smaller than the read length!")

  begin_number <- seq(1,alignment_length-200, by=read_length-overlap); if (verbose) { cat("Beginning position of each subset","\n"); cat(begin_number,"\n")}
  end_number <- begin_number+read_length-1; if (verbose) { cat("Ending position of each subset\n"); cat(end_number,"\n")}
  number_of_subsets <- length(begin_number); if (verbose) { cat(paste0("Total number of subsets = ",number_of_subsets,"\n"))}

  for (i in begin_number) {
    subset_original <- lapply(1:total_reads, function(x) {substr(NGSreads[x],i,i+read_length-1)}) # begin to subdivide the big alignment into subsets, each has the length of read_length
    seqinr::write.fasta(sequences = subset_original, names = 1:length(subset_original), file.out = paste0("Subset_",which(begin_number==i),"_original_size.fasta"))
    subset_small <- subset_original[which(as.character(lapply(1:total_reads, function(x) {substr(subset_original[x],1,10)}))!="----------")]
    subset_smaller <- subset_small[which(as.character(lapply(1:length(subset_small), function(x) {substr(subset_small[x],200,209)}))!="----------")]
    subset_smallest <- subset_smaller[which(lapply(1:total_reads, function(x) {stringr::str_count(substr(subset_smaller[x],1,200),"A")>25})==TRUE)]
    if (verbose) { cat(paste0("Number of reads in downsized subset ", which(begin_number==i), " = ",length(subset_smallest),"\n"))}
    seqinr::write.fasta(sequences = subset_smallest, names = 1:length(subset_smallest), file.out = paste0("Subset_",which(begin_number==i),"_downsized.fasta"))
  }

  cat("Subsetting and downsizing finished!\n")
  beepr::beep(sound = 1, expr = NULL) # make a sound when run finishes
}




R/copy_separate.R
#' @title copy_separate
#'
#' @description Separates two or more gene copies from short-read Next-Generation Sequencing data into a small number of overlapping DNA sequences.
#'
#' @param filename A fasta file contains thousands of short reads that have been mapped to a reference. The reference and reads that are not directly mapped to the reference need to be removed after mapping.
#'
#' @param copy_number An integer (e.g. 2,3, or 4) giving the anticipated number of gene copies in the input file.
#'
#' @param read_length An integer (e.g. 250, or 300) giving the read length of your Next-generation Sequencing data. This method is designed for read length >=250bp.
#'
#' @param overlap An integer describing number of base pairs of overlap between adjacent subsets. More overlap means more subsets. Default 225.
#'
#' @param verbose Turn on (verbose=1; default) or turn off (verbose=0) the output.
#'
#' @return A fasta alignment of a small number of overlapping DNA sequences covering the entire length of the target gene. Gene copies can be assembled by reordering the alignment manually or use the function "copy_assemble".
#'
#' @importFrom seqinr read.fasta write.fasta
#'
#' @importFrom stringr str_count str_sort
#'
#' @importFrom kmer otu
#'
#' @importFrom Biostrings readDNAStringSet
#'
#' @importFrom ape as.character.DNAbin read.FASTA
#'
#' @importFrom DECIPHER ConsensusSequence
#'
#' @importFrom beepr beep
#'
#' @export copy_separate
#'

copy_separate<-function(filename,copy_number,read_length,overlap=225, verbose=1)
{
sink("log.txt", append=FALSE, split=TRUE) # begin to record log
error_log_function <- function() {
  cat(geterrmessage(), file="Error_log.txt", append=T)
}

if (copy_number<=1) stop ("The anticipated copy number must be a number larger than one!")

if (read_length<250) warning ("This method is designed for read length 250bp or longer. Short reads can easily result in chimeric sequences.")

NGSreads <- seqinr::read.fasta(file = filename,seqtype = "DNA", as.string = TRUE,forceDNAtolower = FALSE,set.attributes = FALSE)

total_reads <- length(NGSreads); if (verbose) { cat(paste0("Total number of reads imported = ",total_reads,"\n"))}
alignment_length <- nchar(NGSreads[[1]]); if (verbose) { cat(paste0("Length of the alignment = ",alignment_length,"\n"))}

if (verbose) { cat(paste0("Read length = ",read_length,"\n"))}
if (verbose) { cat(paste0("Overlap between adjacent subsets = ",overlap,"\n"))}
if (overlap>=read_length) stop("Overlap between adjacent subsets must be smaller than the read length!")

begin_number <- seq(1,alignment_length-200, by=read_length-overlap); if (verbose) { cat("Beginning position of each subset","\n"); cat(begin_number,"\n")}
end_number <- begin_number+read_length-1; if (verbose) { cat("Ending position of each subset\n"); cat(end_number,"\n")}
number_of_subsets <- length(begin_number); if (verbose) { cat(paste0("Total number of subsets = ",number_of_subsets,"\n"))}

for (i in begin_number) {
  subset_original <- lapply(1:total_reads, function(x) {substr(NGSreads[x],i,i+read_length-1)}) # begin to subdivide the big alignment into subsets, each has the length of read_length
  subset_small <- subset_original[which(as.character(lapply(1:total_reads, function(x) {substr(subset_original[x],1,10)}))!="----------")]
  subset_smaller <- subset_small[which(as.character(lapply(1:length(subset_small), function(x) {substr(subset_small[x],200,209)}))!="----------")]
  subset_smallest <- subset_smaller[which(lapply(1:total_reads, function(x) {stringr::str_count(substr(subset_smaller[x],1,200),"A")>25})==TRUE)]
  if (verbose) { cat(paste0("Number of reads in subset ", which(begin_number==i), " = ",length(subset_smallest),"\n"))}
  if (length(subset_smallest)<=2) stop("This subset has too few reads (<2). Enter a new value for the parameter 'overlap' and subset the dataset again!")
  seqinr::write.fasta(sequences = subset_smallest, names = 1:length(subset_smallest), file.out = paste0("Subset_",which(begin_number==i),"_downsized.fasta"))
}

Subsets <- stringr::str_sort(list.files(pattern="_downsized.fasta"), numeric = TRUE)

# for each of the subset_1, 2, 3...
for (i in 1:length(Subsets)) {
  if (verbose) { cat("********************************************\n")}
  Sub_set <- ape::as.character.DNAbin(ape::read.FASTA(file=Subsets[i], type = "DNA"))
  if (verbose) { cat((paste0("Clustering analyses for the Subset ", i,"\n")))}

  # find the threshold range for OTU to find the major clusters (number=copy_number) for each subset
  for (m in seq(0.3,1, by = 0.1)) {
    Subset_OTU <- kmer::otu(Sub_set, k = 5, threshold = m, method = "central", nstart = 20)
    if (verbose) { cat(paste0("threshold = ",m),"\n")}
    if (verbose) { cat(unique(Subset_OTU),"\n")}
    if (length(unique(Subset_OTU))>=copy_number) {break}
  }

  # try different threshold values in the range found above
  for (j in seq(m-0.09,m, by = 0.01)) {
    Subset_OTU <- kmer::otu(Sub_set, k = 5, threshold = j, method = "central", nstart = 20)
    if (verbose) { cat(paste0("threshold = ",j),"\n")}
    if (verbose) { cat( unique(Subset_OTU),"\n")}
    if (length(unique(Subset_OTU))>=copy_number) {break}
  }

  reads_each_cluster <- sapply(unique(Subset_OTU), function(x) length(which(Subset_OTU==x)))

  if (verbose) { cat(paste0("Best threshold found = ",j),"\n")}
  if (verbose) { cat(unique(Subset_OTU),"\n")}
  if (verbose) { cat("Number of reads in each cluster\n")}
  if (verbose) { cat(reads_each_cluster,"\n")}

  for (l in (1:copy_number)) {
    Picked_cluster <- Sub_set[which(Subset_OTU==unique(Subset_OTU)[which(reads_each_cluster==sort(reads_each_cluster)[length(unique(Subset_OTU))-l+1])])]
    seqinr::write.fasta(sequences = Picked_cluster, names = labels(Picked_cluster), file.out = paste0("Subset_",i,"_cluster_",l,".fasta"))
    if (verbose) { cat(paste0("Number of reads in picked cluster ",l, " = ", length(Picked_cluster),"\n"))}

    # calcuate the consensus sequence for the clusters of the subset
    seqinr::write.fasta(sequences = paste0(c(as.character(rep("-",(read_length-overlap)*(i-1))),as.character(DECIPHER::ConsensusSequence(Biostrings::readDNAStringSet(paste0("Subset_",i,"_cluster_",l,".fasta"),
                                                                                                                                        format="fasta",nrec=-1L, skip=0L),threshold = 0.4, ambiguity = TRUE, noConsensusChar = "N")[1])), collapse=''),
                names = paste0("Subset_",i,"_cluster_",l,"_consensus"), file.out = paste0("Subset_",i,"_cluster_",l,"_consensus.fasta"))
  }
}

# put together all the consensus sequences from all subsets into one file
Consensus_list <- stringr::str_sort(list.files(pattern="_consensus.fasta"), numeric = TRUE)
All_consensus <- lapply(1:length(Consensus_list), function (x) seqinr::read.fasta(file = Consensus_list[x], seqtype = "DNA",
                                                                          as.string = TRUE,forceDNAtolower = FALSE,set.attributes = FALSE, whole.header = TRUE))
filename_short <- gsub("[:.:].*","", filename) # remove file extensions, e.g. ".fasta", ".txt"

# move subset files into the intermediate files folder
dir.create(paste0(filename_short,"_intermediate_files"))
invisible(file.copy(list.files(pattern="Subset_"), paste0(filename_short,"_intermediate_files")))  # use "invisible" so that output do not show here
unlink("Subset_*")

seqinr::write.fasta(sequences=All_consensus, names=Consensus_list, file.out=paste0(filename_short,"_combined_consensus_",copy_number,"copies_overlap",overlap,".txt"))
cat("Run finished!\n")
beepr::beep(sound = 1, expr = NULL) # make a sound when run finishes
options("error" = error_log_function)
sink() # turn off log
}

